File: tests/generator.test.ts
import { describe, expect, it } from "vitest";
import { MAX_INTERVALS } from "../lib/defaults";
import { calculateMaxInterval, generator } from "../lib/generator";
import { PERIOD, type YearlyPayment } from "../lib/types";
import { createDate } from "../lib/utils";

describe("Generator deletion modifications", () => {
	it("should handle delete future with rest flag", () => {
		const entry = {
			id: "1",
			date: createDate("2024-01-01"),
			config: {
				start: createDate("2024-01-01"),
				period: PERIOD.MONTH,
				interval: 3,
				options: {
					every: 1,
					each: [1, 15],
				},
			},
		};

		const modifications = [
			{
				itemId: "1",
				index: 2,
				payload: { deleted: true },
				restPayload: { deleted: true },
			},
		];

		const result = generator({
			data: [entry],
			modifications,
		});

		// Should only include occurrences before the deletion point
		expect(result).toHaveLength(1);
	});
});

describe("Generator yearly period cases", () => {
	it("should handle yearly payments with ordinal dates", () => {
		const entry = {
			id: "1",
			date: createDate("2024-01-01"),
			config: {
				start: createDate("2024-01-01"),
				period: PERIOD.YEAR,
				interval: 2,
				options: {
					every: 1,
					each: [6], // June
					on: "third-wednesday", // Third Wednesday
				},
			} satisfies YearlyPayment,
		};

		const result = generator({
			data: [entry],
			modifications: [],
		});

		expect(result).toHaveLength(2);
		expect(result[0].actualDate.toString()).toBe("2024-06-19"); // Third Wednesday of June 2024
		expect(result[1].actualDate.toString()).toBe("2025-06-18"); // Third Wednesday of June 2025
	});

	it("should handle invalid ordinal dates in yearly payments", () => {
		const entry = {
			id: "1",
			date: createDate("2024-01-01"),
			config: {
				start: createDate("2024-01-01"),
				period: PERIOD.YEAR,
				interval: 1,
				options: {
					every: 1,
					each: [2], // February
					on: "fifth-wednesday", // Fifth Wednesday (not exist)
				},
			} satisfies YearlyPayment,
		};

		const result = generator({
			data: [entry],
			modifications: [],
		});

		expect(result).toHaveLength(0); // No valid dates found
	});

	it("should handle yearly payments with only ordinal dates (no each)", () => {
		const entry = {
			id: "1",
			date: createDate("2024-01-01"),
			config: {
				start: createDate("2024-01-01"),
				period: PERIOD.YEAR,
				interval: 2,
				options: {
					every: 1,
					on: "first-monday", // First Monday of the year
				},
			} satisfies YearlyPayment,
		};

		const result = generator({
			data: [entry],
			modifications: [],
		});

		expect(result).toHaveLength(2);
		expect(result[0].actualDate.toString()).toBe("2024-01-01"); // First Monday of 2024
		expect(result[1].actualDate.toString()).toBe("2025-01-06"); // First Monday of 2025
	});

	it("should handle non-existent ordinal dates without each option", () => {
		const entry = {
			id: "1",
			date: createDate("2024-01-01"),
			config: {
				start: createDate("2024-01-01"),
				period: PERIOD.YEAR,
				interval: 1,
				options: {
					every: 1,
					on: "fifth-sunday", // Fifth Sunday (may not exist in some months)
				},
			} satisfies YearlyPayment,
		};

		const result = generator({
			data: [entry],
			modifications: [],
		});

		expect(result).toHaveLength(0); // No valid dates found
	});

	it("should handle yearly payments with both each and ordinal dates", () => {
		const entry = {
			id: "1",
			date: createDate("2024-01-01"),
			config: {
				start: createDate("2024-01-01"),
				period: PERIOD.YEAR,
				interval: 1,
				options: {
					every: 1,
					each: [3, 9], // March and September
					on: "first-monday", // First Monday of selected months
				},
			} satisfies YearlyPayment,
		};

		const result = generator({
			data: [entry],
			modifications: [],
		});

		expect(result).toHaveLength(2);
		expect(result[0].actualDate.toString()).toBe("2024-03-04"); // First Monday of March 2024
		expect(result[1].actualDate.toString()).toBe("2024-09-02"); // First Monday of September 2024
	});

	it("should handle yearly payments with invalid ordinal dates in specific months", () => {
		const entry = {
			id: "1",
			date: createDate("2024-01-01"),
			config: {
				start: createDate("2024-01-01"),
				period: PERIOD.YEAR,
				interval: 1,
				options: {
					every: 1,
					each: [2, 6], // February and June
					on: "fifth-thursday", // Fifth Thursday (may not exist in some months)
				},
			} satisfies YearlyPayment,
		};

		const result = generator({
			data: [entry],
			modifications: [],
		});

		expect(result).toHaveLength(1); // Only February 2024 has a fifth Thursday
		expect(result[0].actualDate.toString()).toBe("2024-02-29");
	});

	it("should handle yearly payments where ordinal date becomes invalid after month adjustment", () => {
		const entry = {
			id: "1",
			date: createDate("2024-01-01"),
			config: {
				start: createDate("2024-01-01"),
				period: PERIOD.YEAR,
				interval: 2,
				options: {
					every: 1,
					each: [2, 4], // February and April
					on: "fifth-monday", // Fifth Monday (targeting impossible dates)
				},
			} satisfies YearlyPayment,
		};

		const result = generator({
			data: [entry],
			modifications: [],
			weekendDays: [6, 7], // Saturday and Sunday as weekend days
		});

		// April 2024 has a fifth Monday
		expect(result).toHaveLength(1);
	});
});

describe("Generator modification cases", () => {
	it("should apply future modifications correctly", () => {
		const entry = {
			id: "1",
			date: createDate("2024-01-01"),
			amount: 100,
			config: {
				start: createDate("2024-01-01"),
				period: PERIOD.MONTH,
				interval: 3,
				options: {
					every: 1,
					each: [1],
				},
			},
		};

		const modifications = [
			{
				itemId: "1",
				index: 2,
				payload: { amount: 200 },
				restPayload: { amount: 200 },
			},
		];

		const result = generator({
			data: [entry],
			modifications,
		});

		expect(result).toHaveLength(3);
		expect(result[0].$.amount).toBe(100); // First occurrence unchanged
		expect(result[1].$.amount).toBe(200); // Second occurrence modified
		expect(result[2].$.amount).toBe(200); // Third occurrence also modified due to applyToFuture
	});

	it("should handle multiple deletions with continue and break scenarios", () => {
		const entry = {
			id: "1",
			date: createDate("2024-01-01"),
			config: {
				start: createDate("2024-01-01"),
				period: PERIOD.MONTH,
				interval: 4,
				options: {
					every: 1,
					each: [1, 15], // 1st and 15th of each month
				},
			},
		};

		// this will generate
		// 2024-01-01
		// 2024-01-15 - delete this
		// 2024-02-01
		// 2024-02-15 - delete this and future
		// 2024-03-01 - deleted by future
		// 2024-03-15 - deleted by future
		// 2024-04-01 - deleted by future
		// 2024-04-15 - deleted by future

		const modifications = [
			{
				itemId: "1",
				index: 2,
				payload: { deleted: true },
			},
			{
				itemId: "1",
				index: 4,
				payload: { deleted: true },
				restPayload: { deleted: true },
			},
		];

		const result = generator({
			data: [entry],
			modifications,
		});

		expect(result).toHaveLength(2);
		expect(result.map((r) => r.index)).toEqual([1, 3]); // Verify specific indices
	});
});

describe("calculateMaxInterval", () => {
	it("should return 1 for period 'none'", () => {
		expect(calculateMaxInterval(5, "none", MAX_INTERVALS)).toBe(1);
		expect(calculateMaxInterval(0, "none", MAX_INTERVALS)).toBe(1);
	});

	it("should return the default max interval when interval is 0", () => {
		expect(calculateMaxInterval(0, "week", MAX_INTERVALS)).toBe(
			MAX_INTERVALS.week,
		);
		expect(calculateMaxInterval(0, "month", MAX_INTERVALS)).toBe(
			MAX_INTERVALS.month,
		);
		expect(calculateMaxInterval(0, "year", MAX_INTERVALS)).toBe(
			MAX_INTERVALS.year,
		);
	});

	it("should return the provided interval if it's less than the default max", () => {
		// Assuming MAX_INTERVALS.week > 10, .month > 5, .year > 2 for these assertions to be meaningful
		expect(calculateMaxInterval(10, "week", MAX_INTERVALS)).toBe(10);
		expect(calculateMaxInterval(5, "month", MAX_INTERVALS)).toBe(5);
		expect(calculateMaxInterval(2, "year", MAX_INTERVALS)).toBe(2);
	});

	it("should return the default max interval if the provided interval is greater", () => {
		const highIntervalWeek = MAX_INTERVALS.week + 100;
		const highIntervalMonth = MAX_INTERVALS.month + 100;
		const highIntervalYear = MAX_INTERVALS.year + 100;

		expect(calculateMaxInterval(highIntervalWeek, "week", MAX_INTERVALS)).toBe(
			MAX_INTERVALS.week,
		);
		expect(
			calculateMaxInterval(highIntervalMonth, "month", MAX_INTERVALS),
		).toBe(MAX_INTERVALS.month);
		expect(calculateMaxInterval(highIntervalYear, "year", MAX_INTERVALS)).toBe(
			MAX_INTERVALS.year,
		);
	});

	it("should handle different MAX_INTERVALS configurations", () => {
		const customMaxIntervals = {
			week: 5,
			month: 3,
			year: 1,
		};
		// Interval = 0
		expect(calculateMaxInterval(0, "week", customMaxIntervals)).toBe(
			customMaxIntervals.week,
		);
		expect(calculateMaxInterval(0, "month", customMaxIntervals)).toBe(
			customMaxIntervals.month,
		);
		expect(calculateMaxInterval(0, "year", customMaxIntervals)).toBe(
			customMaxIntervals.year,
		);

		// Interval < default
		expect(calculateMaxInterval(2, "week", customMaxIntervals)).toBe(2);
		expect(calculateMaxInterval(1, "month", customMaxIntervals)).toBe(1);

		// Interval > default
		expect(calculateMaxInterval(10, "week", customMaxIntervals)).toBe(
			customMaxIntervals.week,
		);
		expect(calculateMaxInterval(5, "month", customMaxIntervals)).toBe(
			customMaxIntervals.month,
		);
		expect(calculateMaxInterval(2, "year", customMaxIntervals)).toBe(
			customMaxIntervals.year,
		);
	});
});

describe("Generator weekly period specific cases", () => {
	it("should apply date adjustments correctly for weekly entries without 'each'", () => {
		const entry = {
			id: "weekly-no-each",
			date: createDate("2024-07-01"), // Monday
			amount: 100,
			config: {
				start: createDate("2024-07-01"), // Starts on a Monday
				period: PERIOD.WEEK,
				interval: 4, // Generate 4 weeks
				options: {
					every: 1, // Every week
					// No 'each' specified
				},
			},
		};

		const modifications = [
			{
				itemId: "weekly-no-each",
				index: 2, // Modify the second occurrence
				payload: {
					date: createDate("2024-07-10"), // Change date from Mon July 8 to Wed July 10
				},
				restPayload: {
					// Apply this change to future occurrences as well
					amount: 200, // Include some restPayload to trigger adjustment logic
				},
			},
		];

		const result = generator({
			data: [entry],
			modifications,
		});

		// Expected dates:
		// 1. 2024-07-01 (Original Monday)
		// 2. 2024-07-10 (Modified to Wednesday) -> dateAdjustment = +2 days
		// 3. 2024-07-15 (Original Monday) + 2 days = 2024-07-17 (Wednesday)
		// 4. 2024-07-22 (Original Monday) + 2 days = 2024-07-24 (Wednesday)

		expect(result).toHaveLength(4);
		expect(result[0].actualDate.toString()).toBe("2024-07-01");
		expect(result[0].$.amount).toBe(100);

		expect(result[1].actualDate.toString()).toBe("2024-07-10"); // Modified date
		expect(result[1].$.amount).toBe(200); // Rest payload applied

		// Check subsequent dates reflect the adjustment
		expect(result[2].actualDate.toString()).toBe("2024-07-17"); // Adjusted date (Original Mon + 2 days)
		expect(result[2].$.amount).toBe(200); // Rest payload applied

		expect(result[3].actualDate.toString()).toBe("2024-07-24"); // Adjusted date (Original Mon + 2 days)
		expect(result[3].$.amount).toBe(200); // Rest payload applied
	});
});

File: client-go/generator_test.go
package recurrentry

import (
	"testing"
	"time"
)

func TestSinglePayment(t *testing.T) {
	// Create a start date
	startDate := time.Date(2023, 1, 1, 0, 0, 0, 0, time.UTC)

	// Create a single payment
	singlePayment := BaseEntry{
		ID:   1,
		Date: startDate,
		// No config for a single payment, or period "none"
	}

	// Generate entries
	entries, err := Generator(GeneratorOptions{
		Data: []BaseEntry{singlePayment},
	})

	// Check for errors
	if err != nil {
		t.Fatalf("Error generating entries: %v", err)
	}

	// Check that exactly one entry was generated
	if len(entries) != 1 {
		t.Errorf("Expected 1 entry, got %d", len(entries))
	}

	// Check that the entry has the correct date
	if !entries[0].ActualDate.Equal(startDate) {
		t.Errorf("Expected date %v, got %v", startDate, entries[0].ActualDate)
	}

	// Check that payment date equals actual date for single payment
	if !entries[0].PaymentDate.Equal(entries[0].ActualDate) {
		t.Errorf("Expected payment date to equal actual date")
	}

	// Check that index is 1
	if entries[0].Index != 1 {
		t.Errorf("Expected index 1, got %d", entries[0].Index)
	}
}

func TestMonthlyPayment(t *testing.T) {
	// Create a start date
	startDate := time.Date(2023, 1, 1, 0, 0, 0, 0, time.UTC)

	// Weekend days (6 = Saturday, 7 = Sunday)
	weekendDays := []int{6, 7}

	// Create a monthly payment on the 15th of each month for 3 months
	monthlyPayment := BaseEntry{
		ID:   1,
		Date: startDate,
		Config: &RecurrenceConfig{
			BaseRecurrenceConfig: BaseRecurrenceConfig{
				Start:    startDate,
				Interval: 3, // Generate 3 occurrences
			},
			Period: PeriodMonth,
			Options: RecurrenceOptions{
				WorkdaysOnly: true,
				Every:        1,         // Every month
				Each:         []int{15}, // On the 15th of each month
			},
		},
	}

	// Generate entries
	entries, err := Generator(GeneratorOptions{
		Data:        []BaseEntry{monthlyPayment},
		WeekendDays: weekendDays,
	})

	// Check for errors
	if err != nil {
		t.Fatalf("Error generating entries: %v", err)
	}

	// Check that exactly three entries were generated
	if len(entries) != 3 {
		t.Errorf("Expected 3 entries, got %d", len(entries))
	}

	// Check that the entries have the correct dates
	expectedDates := []time.Time{
		time.Date(2023, 1, 15, 0, 0, 0, 0, time.UTC),
		time.Date(2023, 2, 15, 0, 0, 0, 0, time.UTC),
		time.Date(2023, 3, 15, 0, 0, 0, 0, time.UTC),
	}

	for i, entry := range entries {
		if !entry.ActualDate.Equal(expectedDates[i]) {
			t.Errorf("Entry %d: Expected date %v, got %v", i+1, expectedDates[i], entry.ActualDate)
		}

		// Check that index is correct (1-based)
		if entry.Index != i+1 {
			t.Errorf("Entry %d: Expected index %d, got %d", i+1, i+1, entry.Index)
		}
	}
}

func TestModifications(t *testing.T) {
	// Create a start date
	startDate := time.Date(2023, 1, 1, 0, 0, 0, 0, time.UTC)

	// Create a weekly payment for 5 weeks
	weeklyPayment := BaseEntry{
		ID:   1,
		Date: startDate,
		Config: &RecurrenceConfig{
			BaseRecurrenceConfig: BaseRecurrenceConfig{
				Start:    startDate,
				Interval: 5, // Generate 5 occurrences
			},
			Period: PeriodWeek,
			Options: RecurrenceOptions{
				Every: 1, // Every week
			},
		},
	}

	// Create modifications to specific occurrences
	modifications := []Modification{
		{
			ItemID: 1,
			Index:  3, // Delete the 3rd occurrence
			Payload: map[string]interface{}{
				"deleted": true,
			},
		},
		{
			ItemID: 1,
			Index:  4, // Delete the 4th and all future occurrences
			Payload: map[string]interface{}{
				"deleted": true,
			},
			RestPayload: map[string]interface{}{
				"deleted": true,
			},
		},
	}

	// Generate entries
	entries, err := Generator(GeneratorOptions{
		Data:          []BaseEntry{weeklyPayment},
		Modifications: modifications,
	})

	// Check for errors
	if err != nil {
		t.Fatalf("Error generating entries: %v", err)
	}

	// Check that exactly 2 entries were generated (1st and 2nd, 3rd was deleted, 4th and 5th were deleted via RestPayload)
	if len(entries) != 2 {
		t.Errorf("Expected 2 entries after modifications, got %d", len(entries))
	}

	// Check that the entries have the correct indices
	if entries[0].Index != 1 {
		t.Errorf("Expected first entry to have index 1, got %d", entries[0].Index)
	}
	if entries[1].Index != 2 {
		t.Errorf("Expected second entry to have index 2, got %d", entries[1].Index)
	}
}

func TestGenerator(t *testing.T) {
	baseDate := time.Date(2023, 1, 1, 0, 0, 0, 0, time.UTC)
	weekendDays := []int{6, 7} // Saturday, Sunday
	holidays := []time.Time{
		time.Date(2023, 1, 16, 0, 0, 0, 0, time.UTC), // Martin Luther King Jr. Day
	}

	tests := []struct {
		name        string
		options     GeneratorOptions
		expectedLen int
		expectError bool
	}{
		{
			name: "Simple monthly payment for 3 months",
			options: GeneratorOptions{
				Data: []BaseEntry{
					{
						ID:   1,
						Date: baseDate,
						Config: &RecurrenceConfig{
							BaseRecurrenceConfig: BaseRecurrenceConfig{
								Start:    baseDate,
								Interval: 3, // Generate 3 occurrences
							},
							Period: PeriodMonth,
							Options: RecurrenceOptions{
								Every: 1,         // Every month
								Each:  []int{15}, // On the 15th of each month
							},
						},
					},
				},
				WeekendDays: weekendDays,
			},
			expectedLen: 3,
			expectError: false,
		},
		{
			name: "Weekly payment with specific days",
			options: GeneratorOptions{
				Data: []BaseEntry{
					{
						ID:   1,
						Date: baseDate,
						Config: &RecurrenceConfig{
							BaseRecurrenceConfig: BaseRecurrenceConfig{
								Start:    baseDate,
								Interval: 4, // Generate 4 occurrences
							},
							Period: PeriodWeek,
							Options: RecurrenceOptions{
								Every: 1,           // Every week
								Each:  []int{1, 5}, // Monday and Friday
							},
						},
					},
				},
				WeekendDays: weekendDays,
			},
			expectedLen: 8, // 4 weeks * 2 days per week
			expectError: false,
		},
		{
			name: "Yearly payment with months",
			options: GeneratorOptions{
				Data: []BaseEntry{
					{
						ID:   1,
						Date: baseDate,
						Config: &RecurrenceConfig{
							BaseRecurrenceConfig: BaseRecurrenceConfig{
								Start:    baseDate,
								Interval: 2, // Generate 2 occurrences
							},
							Period: PeriodYear,
							Options: RecurrenceOptions{
								Every:       1,           // Every year
								Each:        []int{1, 7}, // January and July
								GracePeriod: 15,          // 15th of the month
							},
						},
					},
				},
			},
			expectedLen: 4, // 2 years * 2 months per year
			expectError: false,
		},
		{
			name: "One-time payment (no config)",
			options: GeneratorOptions{
				Data: []BaseEntry{
					{
						ID:   1,
						Date: baseDate,
					},
				},
			},
			expectedLen: 1,
			expectError: false,
		},
		{
			name: "Monthly payment with date range",
			options: GeneratorOptions{
				Data: []BaseEntry{
					{
						ID:   1,
						Date: baseDate,
						Config: &RecurrenceConfig{
							BaseRecurrenceConfig: BaseRecurrenceConfig{
								Start:    baseDate,
								Interval: 12, // Generate up to 12 occurrences
							},
							Period: PeriodMonth,
							Options: RecurrenceOptions{
								Every: 1,         // Every month
								Each:  []int{15}, // On the 15th of each month
							},
						},
					},
				},
				Range: &DateRange{
					Start: baseDate,
					End:   baseDate.AddDate(0, 3, 0), // Only include 3 months
				},
			},
			expectedLen: 3, // Limited by range
			expectError: false,
		},
		{
			name: "Multiple entries",
			options: GeneratorOptions{
				Data: []BaseEntry{
					{
						ID:   1,
						Date: baseDate,
						Config: &RecurrenceConfig{
							BaseRecurrenceConfig: BaseRecurrenceConfig{
								Start:    baseDate,
								Interval: 3, // Generate 3 occurrences
							},
							Period: PeriodMonth,
							Options: RecurrenceOptions{
								Every: 1,         // Every month
								Each:  []int{15}, // On the 15th of each month
							},
						},
					},
					{
						ID:   2,
						Date: baseDate,
						Config: &RecurrenceConfig{
							BaseRecurrenceConfig: BaseRecurrenceConfig{
								Start:    baseDate,
								Interval: 2, // Generate 2 occurrences
							},
							Period: PeriodWeek,
							Options: RecurrenceOptions{
								Every: 1,        // Every week
								Each:  []int{1}, // Monday
							},
						},
					},
				},
			},
			expectedLen: 5, // 3 monthly + 2 weekly
			expectError: false,
		},
		{
			name: "With modifications",
			options: GeneratorOptions{
				Data: []BaseEntry{
					{
						ID:   1,
						Date: baseDate,
						Config: &RecurrenceConfig{
							BaseRecurrenceConfig: BaseRecurrenceConfig{
								Start:    baseDate,
								Interval: 5, // Generate 5 occurrences
							},
							Period: PeriodMonth,
							Options: RecurrenceOptions{
								Every: 1,         // Every month
								Each:  []int{15}, // On the 15th of each month
							},
						},
					},
				},
				Modifications: []Modification{
					{
						ItemID: 1,
						Index:  2, // Modify the 2nd occurrence
						Payload: map[string]interface{}{
							"amount": 1500, // Add amount field
						},
					},
					{
						ItemID: 1,
						Index:  3, // Delete the 3rd occurrence
						Payload: map[string]interface{}{
							"deleted": true,
						},
					},
				},
			},
			expectedLen: 4, // 5 occurrences - 1 deleted
			expectError: false,
		},
		{
			name: "With rest payload modifications",
			options: GeneratorOptions{
				Data: []BaseEntry{
					{
						ID:   1,
						Date: baseDate,
						Config: &RecurrenceConfig{
							BaseRecurrenceConfig: BaseRecurrenceConfig{
								Start:    baseDate,
								Interval: 5, // Generate 5 occurrences
							},
							Period: PeriodMonth,
							Options: RecurrenceOptions{
								Every: 1,         // Every month
								Each:  []int{15}, // On the 15th of each month
							},
						},
					},
				},
				Modifications: []Modification{
					{
						ItemID: 1,
						Index:  2, // Modify from the 2nd occurrence onwards
						RestPayload: map[string]interface{}{
							"amount": 2000, // Add amount field to all future occurrences
						},
					},
				},
			},
			expectedLen: 5,
			expectError: false,
		},
		{
			name: "Monthly payment with ordinal",
			options: GeneratorOptions{
				Data: []BaseEntry{
					{
						ID:   1,
						Date: baseDate,
						Config: &RecurrenceConfig{
							BaseRecurrenceConfig: BaseRecurrenceConfig{
								Start:    baseDate,
								Interval: 3, // Generate 3 occurrences
							},
							Period: PeriodMonth,
							Options: RecurrenceOptions{
								Every: 1,              // Every month
								On:    "first-monday", // First Monday of each month
							},
						},
					},
				},
				WeekendDays: weekendDays,
			},
			expectedLen: 3,
			expectError: false,
		},
		{
			name: "Invalid entry (no ID)",
			options: GeneratorOptions{
				Data: []BaseEntry{
					{
						ID:   nil, // Missing ID
						Date: baseDate,
						Config: &RecurrenceConfig{
							BaseRecurrenceConfig: BaseRecurrenceConfig{
								Start:    baseDate,
								Interval: 3,
							},
							Period: PeriodMonth,
							Options: RecurrenceOptions{
								Every: 1,
								Each:  []int{15},
							},
						},
					},
				},
			},
			expectedLen: 0,
			expectError: true,
		},
		{
			name: "With holidays and workdays only",
			options: GeneratorOptions{
				Data: []BaseEntry{
					{
						ID:   1,
						Date: baseDate,
						Config: &RecurrenceConfig{
							BaseRecurrenceConfig: BaseRecurrenceConfig{
								Start:    baseDate,
								Interval: 3, // Generate 3 occurrences
							},
							Period: PeriodMonth,
							Options: RecurrenceOptions{
								Every:        1,         // Every month
								Each:         []int{16}, // On the 16th (coincides with MLK day in test)
								WorkdaysOnly: true,      // Adjust for weekends and holidays
							},
						},
					},
				},
				Holidays:    holidays,
				WeekendDays: weekendDays,
			},
			expectedLen: 3,
			expectError: false,
		},
		{
			name: "Every 2 months pattern",
			options: GeneratorOptions{
				Data: []BaseEntry{
					{
						ID:   1,
						Date: baseDate,
						Config: &RecurrenceConfig{
							BaseRecurrenceConfig: BaseRecurrenceConfig{
								Start:    baseDate,
								Interval: 3, // Generate 3 occurrences
							},
							Period: PeriodMonth,
							Options: RecurrenceOptions{
								Every: 2,         // Every 2 months
								Each:  []int{15}, // On the 15th
							},
						},
					},
				},
			},
			expectedLen: 3,
			expectError: false,
		},
		{
			name: "Custom MaxIntervals",
			options: GeneratorOptions{
				Data: []BaseEntry{
					{
						ID:   1,
						Date: baseDate,
						Config: &RecurrenceConfig{
							BaseRecurrenceConfig: BaseRecurrenceConfig{
								Start:    baseDate,
								Interval: 100, // Try to generate 100 occurrences
							},
							Period: PeriodMonth,
							Options: RecurrenceOptions{
								Every: 1,         // Every month
								Each:  []int{15}, // On the 15th
							},
						},
					},
				},
				MaxIntervals: map[Period]int{
					PeriodMonth: 5, // Limit monthly entries to 5 occurrences
				},
			},
			expectedLen: 5, // Limited by MaxIntervals
			expectError: false,
		},
		{
			name: "Invalid ordinal specification",
			options: GeneratorOptions{
				Data: []BaseEntry{
					{
						ID:   1,
						Date: baseDate,
						Config: &RecurrenceConfig{
							BaseRecurrenceConfig: BaseRecurrenceConfig{
								Start:    baseDate,
								Interval: 3,
							},
							Period: PeriodMonth,
							Options: RecurrenceOptions{
								Every: 1,
								On:    "invalid-ordinal", // Invalid ordinal
							},
						},
					},
				},
			},
			expectedLen: 0,
			expectError: true,
		},
		{
			name: "Using weekday category without weekend days",
			options: GeneratorOptions{
				Data: []BaseEntry{
					{
						ID:   1,
						Date: baseDate,
						Config: &RecurrenceConfig{
							BaseRecurrenceConfig: BaseRecurrenceConfig{
								Start:    baseDate,
								Interval: 3,
							},
							Period: PeriodMonth,
							Options: RecurrenceOptions{
								Every: 1,
								On:    "first-weekday", // Requires weekend days
							},
						},
					},
				},
				WeekendDays: []int{}, // No weekend days defined
			},
			expectedLen: 0,
			expectError: true,
		},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			entries, err := Generator(tc.options)

			if tc.expectError {
				if err == nil {
					t.Errorf("Expected error but got nil")
				}
			} else {
				if err != nil {
					t.Errorf("Unexpected error: %v", err)
				}
				if len(entries) != tc.expectedLen {
					t.Errorf("Expected %d entries, got %d", tc.expectedLen, len(entries))
				}
			}
		})
	}
}

func TestDateAdjustment(t *testing.T) {
	baseDate := time.Date(2023, 1, 15, 0, 0, 0, 0, time.UTC)

	tests := []struct {
		name           string
		originalDate   time.Time
		modifiedDate   time.Time
		period         Period
		expectedMonths int
		expectedDays   int
	}{
		{
			name:           "No adjustment (same date)",
			originalDate:   baseDate,
			modifiedDate:   baseDate,
			period:         PeriodMonth,
			expectedMonths: 0,
			expectedDays:   0,
		},
		{
			name:           "Day adjustment only (monthly)",
			originalDate:   baseDate,
			modifiedDate:   time.Date(2023, 1, 20, 0, 0, 0, 0, time.UTC),
			period:         PeriodMonth,
			expectedMonths: 0,
			expectedDays:   5,
		},
		{
			name:           "Month and day adjustment (yearly)",
			originalDate:   baseDate,
			modifiedDate:   time.Date(2023, 3, 20, 0, 0, 0, 0, time.UTC),
			period:         PeriodYear,
			expectedMonths: 2,
			expectedDays:   5,
		},
		{
			name:           "Day of week adjustment (weekly)",
			originalDate:   time.Date(2023, 1, 2, 0, 0, 0, 0, time.UTC), // Monday
			modifiedDate:   time.Date(2023, 1, 6, 0, 0, 0, 0, time.UTC), // Friday
			period:         PeriodWeek,
			expectedMonths: 0,
			expectedDays:   4, // Difference in day of week
		},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			// Access the unexported function through reflections or recreate its logic
			// Since we can't directly test the calculateDateAdjustment function, we'll recreate it
			adj := DateAdjustment{}

			switch tc.period {
			case PeriodYear:
				// For yearly entries, adjust the month and day position
				adj.Months = int(tc.modifiedDate.Month() - tc.originalDate.Month())
				adj.Days = tc.modifiedDate.Day() - tc.originalDate.Day()
			case PeriodMonth:
				// For monthly entries, only adjust the day position
				adj.Days = tc.modifiedDate.Day() - tc.originalDate.Day()
			case PeriodWeek:
				// For weekly entries, calculate the day-of-week difference
				origDayOfWeek := int(tc.originalDate.Weekday()) + 1
				modDayOfWeek := int(tc.modifiedDate.Weekday()) + 1
				adj.Days = modDayOfWeek - origDayOfWeek
			}

			if adj.Months != tc.expectedMonths {
				t.Errorf("Expected Months adjustment %d, got %d", tc.expectedMonths, adj.Months)
			}
			if adj.Days != tc.expectedDays {
				t.Errorf("Expected Days adjustment %d, got %d", tc.expectedDays, adj.Days)
			}
		})
	}
}
